lib/grizzly/firmware_updates/otw_update_runner.ex

```diff
@@ -1,74 +1,655 @@
+
+  # When the upload is complete, we should get dumped back to the bootloader
+  # prompt. If that doesn't happen soon enough, reset the module and continue.
+  def handle_event(:enter, old_state, :upload_complete, data) do
+    log_state_change(old_state, :upload_complete)
+    :alarm_handler.clear_alarm(@in_progress_alarm)
+
+    {:next_state, :upload_complete, data, [{:state_timeout, :timer.seconds(5), :timeout}]}
+  end
+
+  def handle_event(:enter, old_state, :done, data) do
+    :alarm_handler.clear_alarm(@in_progress_alarm)
+    log_state_change(old_state, :done)
+    data = cleanup(data)
+
+    case data.result do
+      :no_upgrade_needed -> report_status({:done, :skipped})
+      :update_successful -> report_status({:done, :success})
+      other -> report_status({:error, other})
+    end
+
+    {:next_state, :done, data, []}
+  end
+
+  # Reset error count on any other state change.
+  def handle_event(:enter, old_state, new_state, data) do
+    log_state_change(old_state, new_state)
+    {:next_state, new_state, %{data | errors: 0}}
+  end
+
+  #
+  # EVENT HANDLER CALLBACKS
+  #
+
+  # The startup delay is finished. Now go to run_prechecks.
+  def handle_event(:state_timeout, :delay_complete, :startup_delay, data) do
+    {:next_state, :run_prechecks, data, [{:next_event, :internal, :zgw_version_check}]}
+  end
+
+  # Query Z/IP Gateway for the current Z-Wave module version and see if we have
+  # an update to run. If we're unable to get an answer from Z/IP Gateway, we'll
+  # increment the error count and repeat this event after a short delay, up to 3
+  # attempts. If we exceed the max attempts, we'll continue on to try to detect
+  # the Gecko bootloader and upload the firmware with the highest version --
+  # after all, some firmware is better than none firmware.
+  #
+  # If we run into an error like :including or :firmware_updating when trying to
+  # get the version, we should delay and try again (these don't count toward the
+  # max attempts to get the version).
+  def handle_event(:internal, :zgw_version_check, :run_prechecks, %{errors: errors} = data)
+      when errors < 3 do
+    case version_from_zipgateway() do
+      {:ok, version} ->
+        Logger.info("[Grizzly] Detected Z-Wave module version: #{version}")
+        # If we got a version, we can check if we need to update.
+        case Enum.find(data.update_specs, &UpdateSpec.applies?(&1, version)) do
+          nil ->
+            {:next_state, :done, %{data | result: :no_update_needed}}
+
+          spec ->
+            {:next_state, :init, %{data | spec: spec}, [next_event_action(:stop_zipgateway)]}
+        end
+
+      {:error, reason} when reason in [:including, :firmware_updating] ->
+        {:keep_state, %{data | errors: 0},
+         [next_event_action(:zgw_version_check, :timer.seconds(30))]}
+
+      {:error, _reason} ->
+        {:keep_state, %{data | errors: data.errors + 1},
+         [next_event_action(:zgw_version_check, :timer.seconds(5))]}
```

Comment:
I don't see `set_alarm` being called for the `@error_alarm` you defined; perhaps we could set it here? Just an idea! 😄 

```suggestion
      {:error, reason} ->
        :alarm_handler.set_alarm({@error_alarm, reason})
        
        {:keep_state, %{data | errors: data.errors + 1},
         [next_event_action(:zgw_version_check, :timer.seconds(5))]}
```